### PR TITLE
Pearson edexcel Page

### PR DESCRIPTION
**Pull request by AdrianJakubcik**

### CODE REVIEW

File: source/js/about.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    $('.toggle-btn').click(toggleShowLongText);
+
+    /**
+     * Function responsible for hiding and showing long texts on expandable boxes
+     */
+    function toggleShowLongText() {
+        let caller = this.parentElement.parentElement;
+
+        let img = caller.parentElement.children[0].children[0];
```
Comment (DiscreteConcrete) [inline source/js/about.js:11]: i don't quite understand what this block is about

File: source/js/about.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+$(document).ready(function() {
+
+    $('.toggle-btn').click(toggleShowLongText);
+
+    /**
+     * Function responsible for hiding and showing long texts on expandable boxes
+     */
+    function toggleShowLongText() {
```
Comment (DiscreteConcrete) [inline source/js/about.js:8]: this whole function is very hard to follow - can you add more explanation?

File: source/js/carousel.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 /* eslint-disable no-use-before-define */
 
 // get the carousel images
-const images = Array.from(document.querySelectorAll('.carousel-item img'));
+const images = Array.from(document.querySelectorAll('.carousel-item img:not([style])'));
```
Comment (DiscreteConcrete) [inline source/js/carousel.js:4]: some magic here - how are we picking the images?

File: source/js/about.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            skip_img_styling = img.getAttribute('style') ? false : true;
+            if (skip_img_styling)
+                img.style = 'height:auto;';
+        }
+
+
+        let textContainer = caller.children[1].children[0];
+        textContainer.id = textContainer.getAttribute('id') ? textContainer.id : 'this';
+
+        let descripContainer = caller.children[0].children[1].children[0];
```
Comment (DiscreteConcrete) [inline source/js/about.js:23]: what des "descrip" mean?

File: source/js/interview.js [hunk 5]
```diff
@@ -185,9 +185,9 @@ $(document).ready(function () {
 
   // Function that's responsible for generating a hash from a string
```
Comment (DiscreteConcrete) [inline source/js/interview.js:186]: (I might have asked in previous PR) what kind of hash is this? for what purpose? there are many different kinds of hashes

File: source/js/modules/Modal.js [hunk 6]
```diff
@@ -17,8 +17,8 @@ class ModalContainer {
 
     this.HtmlText = this._getHtmlText();
     this.rawText = this._getRawText();
-
-    if (!this.HtmlText || !this.rawText) throw new Error('[ModalContainer] >>> ModalContainer.constructor() >>> Object reference resulted in a nullptr');
+    
+    if (!this.HtmlText && !this.rawText) throw new Error('[ModalContainer] >>> ModalContainer.constructor() >>> Object reference resulted in a nullptr');
```
Comment (DiscreteConcrete) [inline source/js/modules/Modal.js:21]: if there was a mistake here, then maybe worth giving a comment explaining the intuition behind what is happening here

File: source/js/modules/Modal.js [hunk 7]
```diff
@@ -108,10 +109,11 @@ class ModalHandler {
 
   /**
      * `Show the modal-box`
+     * @param {Number} time The time for the modal box to slide down
      */
   show() {
-    if (this.sliding) $(this.view.container.parentElement).slideDown(400);
-    else $(this.view.container.parentElement).fadeIn(400);
+    // if (this.sliding) $(this.view.container.parentElement).animate({ height: '100vh' }, time);
```
Comment (DiscreteConcrete) [inline source/js/modules/Modal.js:115]: why do we keep this as comment?

File: source/js/navigation.js [hunk 8]
```diff
@@ -24,7 +24,7 @@ $(document).ready(() => {
     }
 
     // We check here whether the url contains /school/ section
```
Comment (DiscreteConcrete) [inline source/js/navigation.js:26]: hmm isn't this comment outdated?

**Review by DiscreteConcrete — CHANGES_REQUESTED** [review]

File: Extensions/AboutUsParser.js [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+
+var getFiles = ((source, callback) => {
+    glob(source + '/**/*', callback);
+});
+
+
+class Paragraph {
+    /**
+     * 
+     * @param {Object} object
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply Extensions/AboutUsParser.js:28]: It's just for me to know what to pass as the argument

File: Extensions/AboutUsParser.js [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+    constructor(object) {
+        this.title = object.title;
+        this.short = object.text ? object.text.short : object.short;
+        this.long = object.text ? object.text.long : object.long;
+        this.language = object.language;
+        this.folder = object.folder;
+        this.photo = object.photo ? object.photo : null;
+    }
+
+    getDefaultHtmlCode() {
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply Extensions/AboutUsParser.js:39]: Because as things might get added, the html code we would like the parser to produce may change from thing to thing... so that everyone can add html output of their own to fit their needs, but this one I have added is the default one to be used if you would like to produce html just like that one on AboutUs page.

File: Extensions/AboutUsParser.js [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+ * @param {String} string text to be parsed
+ * @param {Function} data callback function holding data
+ * @returns {Array} 
+ */
+function parseString(string) {
+    let sections = string.split(C.separator); // We split the string by our main separator into individual sections
+    let parsedSections = [];
+
+    sections.map(section => { // We iterate through every parsed section
+        let para = section.split('\r\n').filter(sentence => sentence.trim() ? true : false); // Splitting sections into paragraphs by new line and removing empty lines
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply Extensions/AboutUsParser.js:105]: shorthand for paragraph

File: Extensions/AboutUsParser.js [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+
+/** Asynchronously writes a string to a file
+ * 
+ * @param {String} path Path of the file
+ * @param {String} string String to write into the file
+ * @param {Boolean} appendFile 
+ * @returns {void}
+ */
+function writeFile(path, string, appendFile) {
+    try {
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply Extensions/AboutUsParser.js:154]: In case someone ducks up so that they know where and what happened and when and if they go through the code they will figure out where they went wrong.

File: Extensions/AboutUsParser.js [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+        console.error(error)
+        return;
+    }
+}
+
+/** Creates a Json file with our data
+ * 
+ * @param {Array} array Array of `new Paragraph()` objects
+ */
+function saveAsJson(array) {
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply Extensions/AboutUsParser.js:175]: It's not just the paragraphs, the function generally saves whatever string or array or object you give it as Json

File: Extensions/AboutUsParser.js [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+            let string = getStringFromArray(content.data);
+            if(sectionIndex == 0)
+                createComponent(C.ComponentsPath, section.folder, string, content.language, false, nested);
+            else
+                createComponent(C.ComponentsPath, section.folder, string, content.language, true, nested);
+        });
+    });
+}
+
+/**
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply Extensions/AboutUsParser.js:205]: It's deprected for now so it doesnt have any

File: Extensions/AboutUsParser.js [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+ * @returns {Number} count of subdirectories
+ */
+function getDirLength(dir) {
+    dir = Path.normalize(dir);
+    var parse = dir.split('\\'),
+        index = parse.indexOf(Path.parse(dir).base);
+    return Math.floor(parse.length - index);
+}
+
+// Main loop starts here, actual script execution starts here
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply Extensions/AboutUsParser.js:228]: :D

File: Extensions/AboutUsParser.js [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+
+// First we ask user whether he would like to try and parse a custom file
+if (readlineSync.keyInYN("Would you like to try and parse custom files? ")) {
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply Extensions/AboutUsParser.js:231]: Only necessary input is requested. Everything else is automated, it's due to the fact that I made it possible for the parser to also parse different sections other than that of About Us, that's why the input is there.

File: build.js [hunk 17]
```diff
@@ -460,7 +460,9 @@ if (!success)
 	return;
 // second, we prepare all components
 const enComponents = makeComponentDictionary(getDirname() + '/components/{*,en/*}.html')
+//console.log(enComponents);
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply build.js:463]: Debugging

File: components/en/card_contents.html [hunk 18]
```diff
@@ -5,20 +5,20 @@
 :^) 1-footer :::
 More Info
 :^) 1-link :::
-{fill_parents_html}en/school/highschool/index.html
+{{high_school.link}}
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply components/en/card_contents.html:8]: Because it's transpiled using Handlebars, the link `{{high_school.link}}` is a component

File: components/navigation.html [hunk 19]
```diff
@@ -8,13 +8,20 @@
         <div class="left">
             <!-- Search button -->
             <div class="search-container close">
-                <input type="search" name="searchInp" onchange="sendSearch(this.value, 'search.html')" onkeyup="searchSitesWithResult(this.value, 'search_results', false, true),searchInSite(this.value)" id="search_box" />
+                <!-- <input type="search" name="searchInp" id="searchInp" onchange="sendSearch(this.value, 'search.html')" onkeyup="searchSitesWithResult(this.value, 'search_results', false, true),searchInSite(this.value)" id="search_box" /> -->
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply components/navigation.html:11]: The funny thing is, even I don't know 🤣

File: source/css/interview.css [hunk 20]
```diff
@@ -323,7 +322,7 @@
     .card.interview>img {
         user-select: none;
         -webkit-user-drag: none;
-        height: 100%;
+        /* height: 100%; */
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply source/css/interview.css:325]: I left it there for debugging but I can remove it

File: source/js/about.js [hunk 21]
```diff
@@ -0,0 +1,36 @@
+$(document).ready(function() {
```
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply source/js/about.js:1]: Not much it just holds a function for expanding box texts, really it's not rocket science... just look at the name of the function and that's pretty much it

File: source/js/navigation.js (hunk 8 above)
Comment (AdrianJakubcik) [reply source/js/navigation.js:26]: No why should it?

File: source/js/modules/Modal.js (hunk 7 above)
Comment (AdrianJakubcik) [reply source/js/modules/Modal.js:115]: Just in case, it's our second option if the first one ducks up.

File: source/js/modules/Modal.js (hunk 6 above)
Comment (AdrianJakubcik) [reply source/js/modules/Modal.js:21]: It says clearly `Object reference resulted in a nullptr` aka `Object reference not set to an instance of an object.`

File: source/js/interview.js (hunk 5 above)
Comment (AdrianJakubcik) [reply source/js/interview.js:186]: to generate a hash code that can be used to identify certain things.
the sole purpose and function of this function is to just generate a hash from given string

File: source/js/carousel.js (hunk 3 above)
Comment (AdrianJakubcik) [reply source/js/carousel.js:4]: No magic just standard selector technique acquired from css

File: components/navigation.html (hunk 19 above)
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply components/navigation.html:11]: I think it's for debugging

File: build.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (DiscreteConcrete) [reply build.js:470]: that's code pollution. You have two acceptable options:
1. remove them and let people add console logs to where they find them useful
2. create some flag with which you can enable and disable chunks of debugging functionality. probably the most popular library for this is 'debug' - https://www.npmjs.com/package/debug

File: components/navigation.html (hunk 19 above)
[the remark this answers is not in the document]
Comment (DiscreteConcrete) [reply components/navigation.html:11]: if you don't know, then no one will... please remove or add explanation

File: source/js/carousel.js (hunk 3 above)
Comment (DiscreteConcrete) [reply source/js/carousel.js:4]: but why the ":not([style])" - please explain in code.

File: source/js/modules/Modal.js (hunk 6 above)
Comment (DiscreteConcrete) [reply source/js/modules/Modal.js:21]: Sorry, I wasn't clear enough. Why is "!this.HtmlText && !this.rawText" the condition? what are we intuitively checking, why do both of these need to be present? (answer in code please)

File: source/js/modules/Modal.js (hunk 7 above)
Comment (DiscreteConcrete) [reply source/js/modules/Modal.js:115]: please explain in code for future readers please. why second option? what ducks up here? Ie what are the instructions for when to uncomment this?

File: source/js/navigation.js (hunk 8 above)
Comment (DiscreteConcrete) [reply source/js/navigation.js:26]: now you are checking for both school AND summer sections, so the comment is outdated. please update.

**Review by DiscreteConcrete — CHANGES_REQUESTED** [review]

File: source/files/pearson/pearson – kópia.json [hunk 22]
```diff
@@ -0,0 +1,90 @@
+[{
```
Comment (DiscreteConcrete) [inline source/files/pearson/pearson – kópia.json:1]: wait, why do we have a file with "kopia" in its name?

File: Extensions/AboutUsParser.js [hunk 23]
```diff
@@ -0,0 +1,316 @@
...
+ * @returns {Number} count of subdirectories
+ */
+function countSubDirs(dir) {
+    dir = Path.normalize(dir);
+    var parse = dir.split('\\'),
+        index = parse.indexOf(Path.parse(dir).base);
+    return Math.floor(parse.length - index);
+}
+
+function AboutUs() {
```
Comment (DiscreteConcrete) [inline Extensions/AboutUsParser.js:222]: I am really bothered by transpilation needing human attention to interact with console :disappointed: this makes it incredibly hard to automate. how can we replace this?

**Review by DiscreteConcrete — CHANGES_REQUESTED** [review]

File: source/js/modules/Modal.js (hunk 6 above)
Comment (AdrianJakubcik) [reply source/js/modules/Modal.js:21]: No no, sorry it'a misleading but actually I'm checking whether both of them are empty... because we need at least one of them

File: source/files/pearson/pearson – kópia.json (hunk 22 above)
Comment (AdrianJakubcik) [reply source/files/pearson/pearson – kópia.json:1]: It's a backup file in case the original get's rewritten unwantedly, so that I can revert using this backup. It took me a lot of time to edit those json files to get the right pictures there.

File: source/js/modules/Modal.js (hunk 7 above)
Comment (AdrianJakubcik) [reply source/js/modules/Modal.js:115]: Actually it doesn't need to be there, so I removed it

File: Extensions/AboutUsParser.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (AdrianJakubcik) [reply Extensions/AboutUsParser.js]: It was thrown into a function and exported as module... in case someone would like to use it

File: Extensions/AboutUsParser.js (hunk 23 above)
Comment (AdrianJakubcik) [reply Extensions/AboutUsParser.js:233]: No outcome of our discussion yet leaving this open, but it's unnecessary.

File: Extensions/AboutUsParser.js [hunk 24]
```diff
@@ -162,11 +173,11 @@ function writeFile(path, string, appendFile) {
...
 }
 
-/** Creates a Json file with our data
+/** Creates a Json file with our data stringified as Json
  * 
- * @param {Array} array Array of `new Paragraph()` objects
+ * @param {String} string string to be stringified and saved as a json file
  */
-function saveAsJson(array) {
+function saveAsJson(string) {
```
Comment (DiscreteConcrete) [inline Extensions/AboutUsParser.js:180]: wait, this is broken - it's called string but in the function you still use array. I think best would be to just call it "object" or something like that - as the input can be anything that is stringifiable (of which most general version is simply object)

**Review by DiscreteConcrete — CHANGES_REQUESTED** [review]
